### PR TITLE
Use pywin32 more

### PR DESCRIPTION
**Pull request by exarkun**
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3588

I couldn't get rid of as many of these as I wanted but at least I managed to get rid of a few.  More importantly, there's now more direct unit test coverage for what all this code is supposed to be doing.  There will be some follow-up PRs to further simplify things (eg ticket:3581).

### CODE REVIEW

File: src/allmydata/windows/fixups.py [hunk 1]
```diff
@@ -240,3 +192,107 @@ def unmangle(s):
...
+            try:
+                self._stream.flush()
+            except Exception as e:
+                self._complain("%s.flush: %r from %r" % (self.name, e, self._stream))
+                raise
+
+    def write(self, text):
+        try:
+            if self._hConsole is None:
+                if isinstance(text, unicode):
```
Comment (sajith) [inline src/allmydata/windows/fixups.py:268]: Shouldn't this be `if not isinstance(text, unicode)`?

**Review by sajith — CHANGES_REQUESTED** [review]
Looks mostly fine to me.  I have a couple of concerns though:

* License governing `test._win_subprocess` -- it looks like it came from [this gist](https://gist.github.com/vaab/2ad7051fc193167f15f85ef573e54eb9), but it doesn't state a license.  Since it is more than 150 lines of code, we should tread carefully here.

* `typechecks` are failing on CI with complaints about Windows-related modules, because it is not running on Windows:

```
/tmp/venv/bin/tox -e typechecks
[...]
typechecks run-test: commands[0] | mypy src
src/allmydata/test/_win_subprocess.py:7: error: Module 'ctypes' has no attribute 'windll'
src/allmydata/test/_win_subprocess.py:7: error: Module 'ctypes' has no attribute 'WinError'
src/allmydata/windows/fixups.py:6: error: Module 'ctypes' has no attribute 'WINFUNCTYPE'; maybe "CFUNCTYPE" or "PYFUNCTYPE"?
src/allmydata/windows/fixups.py:6: error: Module 'ctypes' has no attribute 'windll'
src/allmydata/windows/fixups.py:6: error: Module 'ctypes' has no attribute 'WinError'
src/allmydata/windows/fixups.py:6: error: Module 'ctypes' has no attribute 'get_last_error'
Found 6 errors in 2 files (checked 286 source files)
ERROR: InvocationError for command /root/project/.tox/typechecks/bin/mypy src (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   typechecks: commands failed
```

I ran `tox -e typechecks` locally on a Windows VM, and it passed there, so I want to say that the modules themselves are probably fine.

```
C:\Users\vagrant\tahoe-lafs>tox -e typechecks
typechecks installed: attrs==20.3.0,Automat==20.2.0,cffi==1.14.4,constantly==15.1.0,cryptography==3.4.1,foolscap @ git+https://github.com/warner/foolscap@291df027d5770f5edd6df01368d
e3e2cf273ccdf,hyperlink==21.0.0,idna==3.1,incremental==17.5.0,mypy==0.800,mypy-extensions==0.4.3,mypy-zope @ git+https://github.com/Shoobx/mypy-zope@ca0fe617f0303e60faf944e02b47a6b2
717c3921,pyasn1==0.4.8,pyasn1-modules==0.2.8,pycparser==2.20,PyHamcrest==2.0.2,pyOpenSSL==20.0.1,semantic-version==2.8.5,service-identity==18.1.0,setuptools-rust==0.11.6,six==1.15.0
,toml==0.10.2,Twisted==20.3.0,typed-ast==1.4.2,typing-extensions==3.7.4.3,zope.event==4.5.0,zope.interface==5.2.0,zope.schema==6.0.1
typechecks run-test-pre: PYTHONHASHSEED='644'
typechecks run-test: commands[0] | mypy src
Success: no issues found in 287 source files
_______________________________ summary ________________________________
  typechecks: commands succeeded
  congratulations :)
``` 

I'm not sure moving `typechecks` step to Windows to address this is in the scope of this PR, or even advisable. Maybe we can get mypy to ignore the two Windows-related modules?  If we merge this PR as-is, `typechecks` is going to be red for future PRs.

File: src/allmydata/test/_win_subprocess.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
+## issue: https://bugs.python.org/issue19264
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/test/_win_subprocess.py:35]: I contacted the author and got them to 3-clause BSD license it.  Updated to reflect this.

File: src/allmydata/windows/fixups.py (hunk 1 above)
Comment (exarkun) [reply src/allmydata/windows/fixups.py:281]: I don't think so.  If _hConsole is None then the data will be written to _stream instead.  _stream requires UTF-8-encoded bytes.

This was pre-existing code that got moved around a bit.  I probably wouldn't have structured things quite this way but I'd also rather not change them around too much in this PR either.

**Comment by sajith** [comment]
I was trying to figure out why `typechecks` is still failing.  Seems that mypy would be happy with

```python
import sys
assert sys.platform == "win32"
```
but _not_ with:
```python
from sys import platform
assert platform == "win32"
```
That was unexpected.

**Comment by sajith** [comment]
Incidentally, I had to add a `basepython = python3` line under `[testenv:typechecks]`; otherwise `tox -e typechecks` would fail on my machine (Debian 10, where `python` is still Python 2.7) with:

```
Collecting mypy
  Using cached mypy-0.720.tar.gz (1.9 MB)
    ERROR: Command errored out with exit status 1:
     command: tahoe-lafs/.tox/typechecks/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-u9dIqC/mypy/setup.py'"'"'; __file__='"'"'/tmp/pip-install-u9dIqC/mypy/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-install-u9dIqC/mypy/pip-egg-info
         cwd: /tmp/pip-install-u9dIqC/mypy/
    Complete output (1 lines):
    ERROR: You need Python 3.5 or later to use mypy.
```

I'm not sure that changing `basepython` for everyone is a good idea.  Maybe leaving a note would be enough!

**Comment by exarkun** [comment]
> That was unexpected.

Ugh yea that's annoying.  Left a comment warning future-us.

**Comment by exarkun** [comment]
> Incidentally, I had to add a basepython = python3 line under [testenv:typechecks]; otherwise tox -e typechecks would fail on my machine (Debian 10, where python is still Python 2.7) with:

I'm not really a big tox user so I'm not sure what the right path is here.  Would this issue make sense as a new ticket?

**Review by sajith — APPROVED** [review]
Extending coverage on `windows/fixups.py` is going to be a painful exercise.  Let us totally ignore that `codecov/patch` has turned red in this case!

**Comment by sajith** [comment]
> I'm not really a big tox user so I'm not sure what the right path is here. Would this issue make sense as a new ticket?

I don't know really.  Python3 will be the default unless the OS is fairly old, like in my case, so this probably is not a big deal.

**Pull request merged**